### PR TITLE
[NAE-1927] Shared Roles

### PR DESCRIPTION
**Pull request by Kovy95**
# Description

- copying of functionality from enterprise edition

Implements [NAE-1927]

## Dependencies

none

### Third party dependencies

No new dependencies were introduced

### Blocking Pull requests

Depends on  https://github.com/netgrif/petriflow/pull/15

## How Has Been This Tested?

manually

### Test Configuration

| Name                | Tested on |
|---------------------| --------- |
| OS                  |    LinuxMint       |
| Runtime             |    Java 11       |
| Dependency Manager  |    Maven 3.8.4       |
| Framework version   |    SpringBoot 2.6.2       |
| Run parameters      |           |
| Other configuration |           |

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes have been checked, personally or remotely, with @...
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have resolved all conflicts with the target branch of the PR
- [ ] I have updated and synced my code with the target branch
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes:
    - [ ] Lint test
    - [ ] Unit tests
    - [ ] Integration tests
- [ ] I have checked my contribution with code analysis tools:
    - [ ] [SonarCloud](https://sonarcloud.io/project/overview?id=netgrif_application-engine)
    - [ ] [Snyk](https://app.snyk.io/org/netgrif/project/ea82b3b8-658d-41f5-89a7-76cd600da01f)
- [ ] I have made corresponding changes to the documentation:
    - [ ] Developer documentation
    - [ ] User Guides
    - [ ] Migration Guides

[NAE-1927]: https://netgrif.atlassian.net/browse/NAE-1927?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/com/netgrif/application/engine/importer/service/Importer.java [hunk 1]
```diff
@@ -1054,22 +1054,33 @@ protected void createRole(Role importRole) {
             throw new IllegalArgumentException("Role ID '" + ProcessRole.ANONYMOUS_ROLE + "' is a reserved identifier, roles with this ID cannot be defined!");
         }
 
+        ProcessRole role = null;
+        if (importRole.isGlobal() == null || !importRole.isGlobal() || (importRole.isGlobal() && processRoleService.findAllByImportId(ProcessRole.GLOBAL + importRole.getId()).isEmpty())) {
```
Comment (renczesstefan) [inline src/main/java/com/netgrif/application/engine/importer/service/Importer.java:1058]: Maybe this if condition could be extracted to a function for better readability.

**Review by renczesstefan — APPROVED** [review]

**Review by mazarijuraj — APPROVED** [review]

File: docs/roles/shared_roles.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            <id>admin_global</id>
+            <logic>
+                <view>true</view>
+                <perform>true</perform>
+            </logic>
+        </roleRef>
+    </transition>
+...
+```
+When importing a Petri net, the importer checks, whether the global role has already existed.
```
Comment (minop) [inline docs/roles/shared_roles.md:35]: whether the global role already exists.

File: docs/roles/shared_roles.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+If not, the importer creates one. If there has been already one, the importer passes it to a the newly created net.
```
Comment (minop) [inline docs/roles/shared_roles.md:36]: If the role already exists, the importer passes it [...]

File: src/main/java/com/netgrif/application/engine/petrinet/domain/roles/ProcessRole.java [hunk 4]
```diff
@@ -26,20 +26,23 @@ public class ProcessRole extends Imported {
...
     private I18nString name;
 
-    @Getter
     @Setter
     private String netId;
 
     @Setter
     private String description;
 
-    @Getter
```
Comment (minop) [inline src/main/java/com/netgrif/application/engine/petrinet/domain/roles/ProcessRole.java:42]: Getters removed. Intended?

File: src/main/java/com/netgrif/application/engine/petrinet/service/ProcessRoleService.java [hunk 5]
```diff
@@ -87,24 +92,27 @@ public void assignRolesToUser(String userId, Set<String> requestedRolesIds, Logg
         Set<ProcessRole> rolesRemovedFromUser = getRolesRemovedFromUser(userOldRoles, requestedRoles);
 
         String idOfPetriNetContainingRole = getPetriNetIdRoleBelongsTo(rolesNewToUser, rolesRemovedFromUser);
-
-        if (idOfPetriNetContainingRole == null)
+        if (!isGlobalFromFirstRole(rolesNewToUser) && !isGlobalFromFirstRole(rolesRemovedFromUser) && idOfPetriNetContainingRole == null) {
```
Comment (minop) [inline src/main/java/com/netgrif/application/engine/petrinet/service/ProcessRoleService.java:95]: Is assignment of multiple roles restricted to the same type? i.e. is it impossible to assign a net role together with a global role in a single operation? If this is allowed, the net role actions should still be run, no?

File: src/main/java/com/netgrif/application/engine/petrinet/domain/roles/ProcessRole.java (hunk 4 above)
Comment (machacjozef) [reply src/main/java/com/netgrif/application/engine/petrinet/domain/roles/ProcessRole.java:42]: @Getter on the whole class

**Review by renczesstefan — APPROVED** [review]

**Pull request merged**